File: Problem2_DiagonalTraverse.py
# Time Complexity : O(mn)
# Space Complexity : O(1)
# Did this code successfully run on Leetcode : Yes

# Three line explanation of solution in plain english:
"""
1. To return the elements in diagonal order, consider 2 directions - upward and downward.
2. To move in any direction, increment or decrement the corresponding row and column to move to the next diagonal element.
3. When the edges are reached, change the direction by moving to the next row or column. In case of corner position,
    change the column to row and row to column to move to the next element.
"""

# Your code here along with comments explaining your approach

def findDiagonalOrder(self, matrix: List[List[int]]) -> List[int]:
    if not matrix or len(matrix) == 0:
        return []

    row_len = len(matrix)
    col_len = len(matrix[0])
    result = []
    row, col = 0, 0

    # Initializing direction as upward
    direction = 1

    while row < row_len and col < col_len:
        result.append(matrix[row][col])

        # Traverse upward
        if direction == 1:

            # Change direction
            if col == col_len - 1:
                direction = -1
                row += 1
            elif row == 0:
                direction = -1
                col += 1

            # No change in direction
            else:
                row -= 1
                col += 1

        # Traverse downward
        else:

            # Change direction
            if row == row_len - 1:
                direction = 1
                col += 1
            elif col == 0:
                direction = 1
                row += 1

            # No change in direction
            else:
                row += 1
                col -= 1

    return result

File: Problem2_DiagonalTraverse.cs
// Approach : By case analysis, consider moving up and down traversing through the matrix with edge cases
// Space Complexity : O(1)
// Time Complexity : O(m*n) where m is num of rows and n is num of cols
// https://leetcode.com/problems/diagonal-traverse/
public class Solution {
    private static readonly int Up = 0, Down = 1;
    
    public int[] FindDiagonalOrder(int[][] mat) {        
        int m = mat.Length, n = mat[0].Length;
        int[] ans = new int[m*n];
        
        int count = 0;
        int direction = Up;
        
        int row=0, col=0;
        
        while(count < m*n){
            ans[count] = mat[row][col];
            count++;
            
            if(direction == Up){
                if(col == n-1){
                    direction = Down;
                    row++;
                } else if(row == 0){
                    direction = Down;
                    col++;
                } else {
                    row--;
                    col++;
                }
                
            } else {
                if(row == m-1){
                    direction = Up;
                    col++;
                } else if(col == 0){
                    direction = Up;
                    row++;
                } else {
                    row++;
                    col--;
                }
            }
            
        }
        
        return ans;
    }
}
